python-sdk/nuscenes/eval/detection/tests/test_algo.py

```diff
@@ -257,5 +281,131 @@ def test_two_imgs(self):
+                 min_recall: float = 0.1):
+        """
+        Calculate and check the AP value.
+        :param gts: Ground truth data.
+        :param preds: Predictions.
+        :param target_error: Expected error value.
+        :param metric_name: Name of the TP metric.
+        :param detection_name: Name of the class we are interested in.
+        :param min_recall: Minimum recall value.
+        """
+
+        metric_data = get_metric_data(gts, preds, detection_name, 2.0)  # Distance threshold for TP metrics is 2.0
+        tp_error = calc_tp(metric_data, min_recall=min_recall, metric_name=metric_name)
+        # We quantize the error curve into 100 bins to calculate the metric so it is only accurate up to 1%.
+        self.assertGreaterEqual(0.01, abs(tp_error - target_error), msg='Incorrect {} value'.format(metric_name))
+
+    def test_no_positives(self):
+        """ Tests the error if there are no matches. The expected behaviour is to return error of 1.0. """
+
+        # Same type of objects but are more than 2m away.
+        car1 = {'trans': (1, 1, 1), 'score': 1.0}
+        car2 = {'trans': (3, 3, 1), 'score': 1.0}
+        bike1 = {'trans': (1, 1, 1), 'score': 1.0, 'name': 'bicycle', 'attr': 'cycle.with_rider'}
+        for metric_name in TP_METRICS:
+            self.check_tp({'sample1': [car1]}, {'sample1': [car2]}, target_error=1.0, metric_name=metric_name)
+
+        # Within distance threshold away but different classes.
+        for metric_name in TP_METRICS:
+            self.check_tp({'sample1': [car1]}, {'sample1': [bike1]}, target_error=1.0, metric_name=metric_name)
+
+    def test_perfect(self):
+        """ Tests when everything is estimated perfectly. """
+
+        car1 = {'trans': (1, 1, 1), 'score': 1.0}
+        car2 = {'trans': (1, 1, 1), 'score': 0.3}
+        for metric_name in TP_METRICS:
+            # Detected with perfect score.
+            self.check_tp({'sample1': [car1]}, {'sample1': [car1]}, target_error=0.0, metric_name=metric_name)
+
+            # Detected with low score.
+            self.check_tp({'sample1': [car1]}, {'sample1': [car2]}, target_error=0.0, metric_name=metric_name)
+
+    def test_one_img(self):
+        """ Test single sample case. """
+
+        # Note all the following unit tests can be repeated to other metrics, but they are not needed.
+        # The intention of these tests is to measure the calc_tp function which is common for all metrics.
+
+        gt1 = {'trans': (1, 1, 1)}
+        gt2 = {'trans': (10, 10, 1), 'size': (2, 2, 2)}
+        gt3 = {'trans': (20, 20, 1), 'size': (2, 4, 2)}
+
+        pred1 = {'trans': (1, 1, 1), 'score': 1.0}
+        pred2 = {'trans': (11, 10, 1), 'size': (2, 2, 2), 'score': 0.9}
+        pred3 = {'trans': (100, 10, 1), 'size': (2, 2, 2), 'score': 0.8}
+        pred4 = {'trans': (20, 20, 1), 'size': (2, 4, 2), 'score': 0.7}
+        pred5 = {'trans': (21, 20, 1), 'size': (2, 4, 2), 'score': 0.7}
+
+        # One GT and one detection
```

Comment:
I think the comment can be more explicit. Like: one GT and one matching prediction. Distance is 1 meter, so error 1. (I'd actually suggest using another value than 1.0 to make it more clear).